sympy/assumptions/ask.py

```diff
@@ -1257,54 +1281,58 @@ def ask(proposition, assumptions=True, context=global_assumptions):
 
+    if local_facts.clauses:
+        local_facts_ = CNF.CNF_to_cnf(local_facts)
+
+        # See if there's a straight-forward conclusion we can make for the inference
+        if local_facts_.is_Atom:
+            if key in known_facts_dict[local_facts_]:
+                return True
+            if Not(key) in known_facts_dict[local_facts_]:
+                return False
+        elif (isinstance(local_facts_, And) and
+              all(k in known_facts_dict for k in local_facts_.args)):
+            for assum in local_facts_.args:
+                if assum.is_Atom:
+                    if key in known_facts_dict[assum]:
+                        return True
+                    if Not(key) in known_facts_dict[assum]:
+                        return False
+                elif isinstance(assum, Not) and assum.args[0].is_Atom:
+                    if key in known_facts_dict[assum]:
+                        return False
+                    if Not(key) in known_facts_dict[assum]:
+                        return True
+        elif (isinstance(key, Predicate) and
+              isinstance(local_facts_, Not) and local_facts_.args[0].is_Atom):
+            if local_facts_.args[0] in known_facts_dict[key]:
+                return False
+
     # direct resolution method, no logic
     res = key(expr)._eval_ask(assumptions)
     if res is not None:
         return bool(res)
 
-    if local_facts is None:
-        return satask(proposition, assumptions=assumptions, context=context)
-
-
-    # See if there's a straight-forward conclusion we can make for the inference
-    if local_facts.is_Atom:
-        if key in known_facts_dict[local_facts]:
-            return True
-        if Not(key) in known_facts_dict[local_facts]:
-            return False
-    elif (isinstance(local_facts, And) and
-            all(k in known_facts_dict for k in local_facts.args)):
-        for assum in local_facts.args:
-            if assum.is_Atom:
-                if key in known_facts_dict[assum]:
-                    return True
-                if Not(key) in known_facts_dict[assum]:
-                    return False
-            elif isinstance(assum, Not) and assum.args[0].is_Atom:
-                if key in known_facts_dict[assum]:
-                    return False
-                if Not(key) in known_facts_dict[assum]:
-                    return True
-    elif (isinstance(key, Predicate) and
-            isinstance(local_facts, Not) and local_facts.args[0].is_Atom):
-        if local_facts.args[0] in known_facts_dict[key]:
-            return False
-
-    # Failing all else, we do a full logical inference
-    res = ask_full_inference(key, local_facts, known_facts_cnf)
-    if res is None:
-        return satask(proposition, assumptions=assumptions, context=context)
+    # using satask (still costly)
+    res = satask(proposition, assumptions=assumptions, context=context)
+
+    if res is not None:
```

Comment:
It this test needed? It looks like `res` will be returned anyway on the next line.

Reply:
Thanks for pointing out. Didn't realize.